source/concurrent/Promise.ooc

```diff
@@ -61,6 +61,30 @@ ClosurePromise: class extends Promise {
 	wait: override func (time: TimeSpan) -> Bool { this _wait(time) }
 }
 
+ConditionPromise: class extends Promise {
+	_completed := false
+	_condition := WaitCondition new()
+	_mutex := Mutex new()
+	init: func { super() }
+	free: override func {
+		(this _mutex, this _condition) free()
+		super()
+	}
+	signal: func {
+		this _mutex with(||
+			this _completed = true
+			this _condition broadcast()
+		)
+	}
+	wait: override func (time: TimeSpan) -> Bool {
```

Comment:
the timed wait should give an error if the argument is not `TimeSpan maximumValue`, but keep all the code you have now.
Then you can always call `wait()` without argument, and everyone is happy.